### DISCUSSION TITLE
Webpack doesn't minify export name on external module end when using `aurelia-bootstrapper` in production mode

### DISCUSSION THREAD

**Discussion by vilnytskyi** [webpack/webpack#17300] [Q&A] [answered]
I'm encountering minification-related issue when bundling aurelia v1 app in production mode: it does not minify some dependencies on both ends, just import side, e.g code from my simple reproduction case:
```
import { ConsoleAppender } from 'aurelia-logging-console'
class CustomConsoleAppender extends ConsoleAppender { }
console.log(new CustomConsoleAppender())
```
bundles into something like
```
var i=n("aurelia-logging-console")
class r extends i.I{}
console.log(new r())
```
but the `i` doesn't contain `I` class, but originally named `ConsoleAppender` instead so I'm receiving undefined as the result.

The interesting thing is that when replacing import statement with
```
const { ConsoleAppender } = require('aurelia-logging-console')
```
it then doesn't minify export name on import side and thus works just fine.
Any ideas what might be wrong? Is it a webpack issue that should be reported?

**Comment by alexander-akait** [comment]
@vilnytskyi I think it is related to `terser`, they have a big problems with `extends` minications, so they may have disabled it temporarily

**Reply by vilnytskyi** [reply]
Are there any alternatives to `terser` that can be plugged in as a replacement? Or maybe I can still use latest `webpack` but explicitly set some older `terser` version that still works?

**Reply by alexander-akait** [reply]
Try to update all deps to the latest and check it again

**Reply by alexander-akait** [reply]
But honestly I would like to investigate this, you can create a minimal repo?

**Reply by vilnytskyi** [reply]
@alexander-akait [Here.](https://codesandbox.io/s/unruffled-http-c72vbg) Just try it locally as it seems that CodeSandbox is still bundling in dev mode or smth, as it works there, which is another interesting fact...

**Reply by alexander-akait** [reply]
@vilnytskyi I done fresh install and everything works perfectly, `terser` makes a release 9 hours ago, can you try to update it? 

Generated code is `console.log(t), console.log(new (class extends t {})());`, also check you have the latest version of webpack (run `npm ls webpack`), we have two problems with externds recently - https://github.com/terser/terser/issues/1392 and https://github.com/webpack/webpack/issues/17264 (sounds like you have it)

And they were fixed

**Reply by vilnytskyi** [reply]
@alexander-akait with latest `webpack@5.85.0` and manually installed `terser@5.17.7` I do still have the same broken bundle. Could you clarify your "fresh install" meaning? What steps have you done after downloading and extracting template from sandbox?

**Reply by alexander-akait** [reply]
@vilnytskyi download your repo and run `npm i`, then run `webpack` to bundle, so let's try:
- run `npm ls webpack` and `npm ls terser` to check that we really have the latest versions
- try to set optimization: `{ minimize: false }`, it will disable terser, but apply all possible optimizations, so if you will have not problems - the problem in terser, if you have, the problem can be in webpack
- try to reproduce your problem using https://codesandbox.io/s/unruffled-http-c72vbg and describe all steps to see the problem :smile:

**Reply by vilnytskyi** [reply]
@alexander-akait followed the steps strictly, all dependencies have latest `webpack` and `terser`, with `minimize: false` issue is still there. Just to confirm, you did use `entry: { app: ["aurelia-bootstrapper"] }` as it is in my repro case right? Because the generated code you mentioned is the same for me only when using `./src/main.ts` as entrypoint but it's not an option for me. `aurelia-bootstrapper` is responsible for preserving some module names so webpack doesn't mangle them into chunk ids, but your result looks like either `aurelia-bootstrapper` didn't affect the resulting bundle or you have changed the entry point...

**Reply by vilnytskyi** [reply]
I would gladly try to reproduce it on codesandbox, but is there an ability to view generated output bundle somehow?

**Reply by alexander-akait** [reply]
@vilnytskyi I see the problem...

**Reply by alexander-akait** [reply]
> with minimize: false issue is still there.

~~Are you sure? Because:~~
```
console.log(aurelia_logging_console__WEBPACK_IMPORTED_MODULE_0__.ConsoleAppender)
```

```
ƒ ConsoleAppender() {
    
  }
```

Reproduced

**Reply by vilnytskyi** [reply]
@alexander-akait this is from my bundle with `minimize: false`:
```
class CustomConsoleAppender extends aurelia_logging_console__WEBPACK_IMPORTED_MODULE_0__/* .ConsoleAppender */ .I {
```

**Reply by alexander-akait** [reply, marked as answer]
`AureliaDependency` prevents module concationation and allow mangling... webpack tries to optimize exports (i.e. `optimizations.mangleExports`) , https://github.com/aurelia/webpack-plugin/blob/master/src/IncludeDependency.ts#L21 is a line where a problem, if you comment it your code will work

**Reply by alexander-akait** [reply]
Based on the provided comment, it should be:
```js
getReferencedExports(moduleGraph) {
        return require("webpack").Dependency.EXPORTS_OBJECT_REFERENCED;;
    }
```    

But maybe I'm missing something, I think it's better to open a topic in https://github.com/aurelia/webpack-plugin/ and ask why it's written this way

**Reply by alexander-akait** [reply]
Try it locally to make sure it fixes all possible places

**Reply by vilnytskyi** [reply]
It works indeed! So is it `webpack` or `aurelia-webpack-loader` issue that is responsible for mangling only import side but not export side? Because I mean they added this line function for some reason, so I have to investigate, if it doesn't break anything in runtime, but still feels strange that it's breaking only one side, but other side is still mangled....

**Reply by vilnytskyi** [reply]
@alexander-akait ok then it seems like it's on their end indeed, thanks a lot for pointing this out!

**Reply by vilnytskyi** [reply]
It indeed fixed all the cases and seems to work correctly in runtime! Thanks once again, I will proceed with MR to the plugin repo then.

**Reply by alexander-akait** [reply]
@vilnytskyi yeah, that is why we need to ask, if we want to mangeExports we need to rewrite logic, if no, we can change a line above and no bugs anymore, but I am seeing `this.options.exports` and I am afraid there is a logic for some cases, unfortunately I don't have enough time to go over how it all works, but I can assist to fix it, you can ping me in the opened issue
